Sources/GRPCOTelTracingInterceptors/Tracing/SpanAttributes+GRPCTracingKeys.swift

```diff
@@ -136,47 +136,91 @@ package enum PeerAddress: Equatable {
         return nil
       }
 
-    case "ipv6":
-      if addressComponents[1].first == "[" {
-        // At this point, we are looking at an address with format: [<address>]:<port>
-        // We drop the first character ('[') and split by ']:' to keep two components: the address
-        // and the port.
-        let ipv6AddressComponents = addressComponents[1].dropFirst().split(separator: "]:")
-        if ipv6AddressComponents.count == 2, let port = Int(ipv6AddressComponents[1]) {
-          self = .ipv6(address: String(ipv6AddressComponents[0]), port: port)
-        } else {
-          return nil
-        }
+      let hostComponent = addressWithoutType[..<addressColon]
+      var portComponent = addressWithoutType[addressColon...]
+      portComponent.removeFirst()
+
+      if let host = String(hostComponent), let port = Int(ipAddressPortStringBytes: portComponent) {
+        self = .ipv4(address: host, port: port)
       } else {
         return nil
       }
+    } else if addressType.elementsEqual("ipv6".utf8) {
+      guard let lastColonIndex = addressWithoutType.lastIndex(of: UInt8(ascii: ":")) else {
+        // This is some unexpected/unknown format
+        return nil
+      }
 
-    case "unix":
-      // Whatever comes after "unix:" is the <pathname>
-      self = .unixDomainSocket(path: String(addressComponents[1]))
+      var hostComponent = addressWithoutType[..<lastColonIndex]
+      var portComponent = addressWithoutType[lastColonIndex...]
+      portComponent.removeFirst()
 
-    default:
+      if let firstBracket = hostComponent.popFirst(), let lastBracket = hostComponent.popLast(),
+        firstBracket == UInt8(ascii: "["), lastBracket == UInt8(ascii: "]"),
+        let host = String(hostComponent), let port = Int(ipAddressPortStringBytes: portComponent)
+      {
+        self = .ipv6(address: host, port: port)
+      } else {
+        // This is some unexpected/unknown format
+        return nil
+      }
+    } else if addressType.elementsEqual("unix".utf8) {
+      // Whatever comes after "unix:" is the <pathname>
+      self = .unixDomainSocket(path: String(addressWithoutType) ?? "")
+    } else {
       // This is some unexpected/unknown format
       return nil
     }
   }
 }
+
+extension Int {
+  package init?(ipAddressPortStringBytes: some Collection<UInt8>) {
+    guard (1 ... 5).contains(ipAddressPortStringBytes.count) else {
+      // Valid IP port values go up to 2^16-1 (65535), which is 5 digits long.
+      // If the string we get is over 5 characters, we know for sure that this is an invalid port.
+      // If it's empty, we also know it's invalid as we need at least one digit.
+      return nil
+    }
+
+    var value = 0
+    for utf8Char in ipAddressPortStringBytes {
+      value &*= 10
+      guard (UInt8(ascii: "0") ... UInt8(ascii: "9")).contains(utf8Char) else {
+        // non-digit character
+        return nil
+      }
+      value &+= Int(utf8Char)
+    }
+
+    guard value <= 65535 else {
```

Comment:
nit: let the compiler do the maths

```suggestion
    guard value <= Int(UInt16.max) else {
```